docs/grammar/SolidityParser.g4

```diff
@@ -311,11 +311,35 @@ errorDefinition:
 	LParen (parameters+=errorParameter (Comma parameters+=errorParameter)*)? RParen
 	Semicolon;
 
+/**
+ * Operators that users are allowed to implement for some types with `using for`.
+ */
+userDefinableOperator:
+	BitAnd
+	| BitNot
+	| BitOr
+	| BitXor
+	| Sar
+	| Shl
+	| Add
+	| Div
+	| Exp
```

Comment:
Built-in versions of `<<`, `>>` and `**` require a signed operand on the right side. This is another thing that was a bit awkward when I tried to implement them for a type that's just a straight wrapper over `uint`:

For example:
```solidity
type T is uint;
using {exp as **} for T;

function exp(T a, T b) pure returns (uint) {
    require(T.unwrap(b) <= type(int).max);
    return T.wrap(T.unwrap(a) ** int(T.unwrap(b)));
}
```

Note that it requires a cast and a check for values that are out of range. Maybe we should allow the types to be different in this case?

Or maybe we should disallow them for now? Note that these operators were not user-definable in the initial implementation but the restriction was relaxed in review: https://github.com/ethereum/solidity/pull/12362#discussion_r953894665.

Reply:
Exponentiation should definitely take an unsigned integer as operand, i.e. be ``'a -> uint -> 'a``.
The dogma is not to have same types everywhere, the dogma is to have only one polymorphic type occuring int he signature. I.e. for *any* exponentiation at *any* type the concrete type `uint` and no other type can be the exponent type.

For shifts it's a bit trickier, let me think a moment there...

Reply:
Similarly to that (I hope at least :-)?) comparisons are not ``'a -> 'a -> 'a``, but fixed ``'a -> 'a -> bool``.

Reply:
Well, if it makes sense to use `uint` for exponentiation then for shifts probably too. They're like division/multiplication by `2^n` and the parameter `n` is also an exponent.

Reply:
So now the questions is - should it be specifically `uint` or any unsigned type?

Reply:
> Similarly to that (I hope at least :-)?) comparisons are not `'a -> 'a -> 'a`, but fixed `'a -> 'a -> bool`.

Yeah, comparisons are `'a -> 'a -> bool`.

Like I said, the original PR did not have these operators and they were added later - without much consideration for parameter types. I wasn't sure if it was a good idea in the first place, but the way you describe it with `uint`s makes sense to me.

Reply:
Hm, yeah, that's indeed the core question... conceptually it has to be a single fixed concrete type... but *also* all builtin exponentiations should be representable as user-defined operators eventually... which kind of means it has to be ``uint256`` specifically... but then implicit conversions for operands come into play...

It's no coincidence that we have weird cases that produce warnings for exponent types generally :-)...

The safest course of action would probably be to postpone all three operators and not rush design decisions on this now.

Reply:
Ultimately it will *probably* be `uint256`. But yeah, maybe better to think through implicit conversions on operands and relate that to the builtin versions of those operators before rushing.

Reply:
ok then. I'll remove them and we can consider that properly for 0.8.19.

Reply:
`<<`, `>>` and `**` disallowed.